lib/Fmt_ast.ml

```diff
@@ -4471,7 +4482,19 @@ let fmt_code ~debug =
         let cmts = Cmts.init Structure ~debug source ast comments in
         let ctx = Pld (PStr ast) in
         Ok (fmt_file ~ctx ~debug Structure source cmts conf ast ~fmt_code)
-    | exception _ -> Error ()
+    | exception _ ->
+        if String.is_prefix (String.strip s) ~prefix:"# " then
+          match
+            Parse_with_comments.parse Parse.ast Repl_file conf ~source:s
+          with
+          | {ast; comments; source; prefix= _} ->
+              let cmts = Cmts.init Repl_file ~debug source ast comments in
+              let ctx = Top in
+              Ok
+                (fmt_file ~ctx ~debug Repl_file source cmts conf ast
+                   ~fmt_code )
```

Comment:
This code is mostly the same as the "structure" case (the only difference are the "kind" and `ctx`). I think it'd be worth having a "is_repl_block" function that can be shared with the normalization (it's important that it's using the exact same).

Reply:
I added a function `Docstring.is_repl_block`

I'm not sure it's possible to factorize these 2 parts of code because of the gadt, I tried a few things but the first instantiation of the fragment makes it a `Structure Extended_ast.t` instead of the generic `a Extended_ast.t`.

Reply:
I've remove some of the duplicated code here: https://github.com/gpetiot/ocamlformat/pull/6
But I also attempted to remove the parsing attempts, let me know what you think.

Reply:
merged, thanks!